File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/SyncBackendList.java
package org.totschnig.myexpenses.fragment;

import android.content.ContentResolver;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ExpandableListView;

import com.dropbox.core.InvalidAccessTokenException;
import com.google.android.material.snackbar.Snackbar;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.BaseActivity;
import org.totschnig.myexpenses.activity.ManageSyncBackends;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.adapter.SyncBackendAdapter;
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment;
import org.totschnig.myexpenses.dialog.DialogUtils;
import org.totschnig.myexpenses.dialog.MessageDialogFragment;
import org.totschnig.myexpenses.model.Account;
import org.totschnig.myexpenses.model.ContribFeature;
import org.totschnig.myexpenses.model.CurrencyContext;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.sync.GenericAccountService;
import org.totschnig.myexpenses.util.UiUtils;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.licence.LicenceHandler;
import org.totschnig.myexpenses.viewmodel.AbstractSyncBackendViewModel;

import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Pair;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import eltos.simpledialogfragment.SimpleDialog;

import static com.google.android.material.snackbar.Snackbar.LENGTH_INDEFINITE;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_SYNC_ACCOUNT_NAME;

public class SyncBackendList extends Fragment implements
    ExpandableListView.OnGroupExpandListener, SimpleDialog.OnDialogResultListener {

  private static final String DIALOG_INACTIVE_BACKEND = "inactive_backend";

  private SyncBackendAdapter syncBackendAdapter;
  private ExpandableListView listView;
  private int metadataLoadingCount = 0;
  private Snackbar snackbar;
  private AbstractSyncBackendViewModel viewModel;

  @Inject
  PrefHandler prefHandler;
  @Inject
  CurrencyContext currencyContext;
  @Inject
  Class<? extends AbstractSyncBackendViewModel> modelClass;
  @Inject
  LicenceHandler licenceHandler;

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    ((MyApplication) requireActivity().getApplication()).getAppComponent().inject(this);
    super.onCreate(savedInstanceState);
    setHasOptionsMenu(true);
    viewModel = new ViewModelProvider(this).get(modelClass);
  }

  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    ProtectedFragmentActivity context = (ProtectedFragmentActivity) getActivity();
    View v = inflater.inflate(R.layout.sync_backends_list, container, false);
    listView = v.findViewById(R.id.list);
    View emptyView = v.findViewById(R.id.empty);
    syncBackendAdapter = new SyncBackendAdapter(context, currencyContext, getAccountList());
    listView.setAdapter(syncBackendAdapter);
    listView.setEmptyView(emptyView);
    listView.setOnGroupExpandListener(this);
    snackbar = Snackbar.make(listView, R.string.sync_loading_accounts_from_backend, LENGTH_INDEFINITE);
    UiUtils.increaseSnackbarMaxLines(snackbar);
    viewModel.getLocalAccountInfo().observe(getViewLifecycleOwner(),
        stringStringMap -> syncBackendAdapter.setLocalAccountInfo(stringStringMap));
    viewModel.loadLocalAccountInfo();
    registerForContextMenu(listView);
    return v;
  }

  @Override
  public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v, ContextMenu.ContextMenuInfo menuInfo) {
    long packedPosition = ((ExpandableListView.ExpandableListContextMenuInfo) menuInfo).packedPosition;
    int commandId;
    int titleId;
    boolean isSyncAvailable = licenceHandler.hasTrialAccessTo(ContribFeature.SYNCHRONIZATION);
    if (ExpandableListView.getPackedPositionType(packedPosition) ==
        ExpandableListView.PACKED_POSITION_TYPE_CHILD) {
      if (isSyncAvailable) {
        switch (syncBackendAdapter.getSyncState(packedPosition)) {
          case SYNCED_TO_THIS:
            commandId = R.id.SYNC_UNLINK_COMMAND;
            titleId = R.string.menu_sync_unlink;
            break;
          case UNSYNCED:
            commandId = R.id.SYNC_LINK_COMMAND;
            titleId = R.string.menu_sync_link;
            break;
          case SYNCED_TO_OTHER:
            commandId = R.id.SYNCED_TO_OTHER_COMMAND;
            titleId = R.string.menu_sync_link;
            break;
          case UNKNOWN:
            commandId = R.id.SYNC_DOWNLOAD_COMMAND;
            titleId = R.string.menu_sync_download;
            break;
          default://ERROR
            commandId = 0;
            titleId = 0;
        }
        if (commandId != 0)
          menu.add(Menu.NONE, commandId, 0, titleId);
      }
    } else {
      if (isSyncAvailable) {
        menu.add(Menu.NONE, R.id.SYNC_COMMAND, 0, R.string.menu_sync_now);
      }
      menu.add(Menu.NONE, R.id.SYNC_REMOVE_BACKEND_COMMAND, 0, R.string.menu_remove);
    }

    super.onCreateContextMenu(menu, v, menuInfo);
  }

  protected List<Pair<String, Boolean>> getAccountList() {
    return viewModel.getAccounts(requireActivity());
  }

  @Override
  public boolean onContextItemSelected(MenuItem item) {
    long packedPosition = ((ExpandableListView.ExpandableListContextMenuInfo) item.getMenuInfo())
        .packedPosition;
    int itemId = item.getItemId();
    ((BaseActivity) requireActivity()).trackCommand(itemId);
    if (itemId == R.id.SYNC_COMMAND) {
      requestSync(packedPosition);
      return true;
    } else if (itemId == R.id.SYNC_UNLINK_COMMAND) {
      final Account accountForSync = getAccountForSync(packedPosition);
      if (accountForSync != null) {
        DialogUtils.showSyncUnlinkConfirmationDialog(requireActivity(),
            accountForSync);
      }
      return true;
    } else if (itemId == R.id.SYNCED_TO_OTHER_COMMAND) {
      Account account = getAccountForSync(packedPosition);
      if (account != null) {
        ((ProtectedFragmentActivity) requireActivity()).showMessage(
            getString(R.string.dialog_synced_to_other, account.getUuid()));
      }
      return true;
    } else if (itemId == R.id.SYNC_LINK_COMMAND) {
      Account account = getAccountForSync(packedPosition);
      if (account != null) {
        MessageDialogFragment.newInstance(
            getString(R.string.menu_sync_link),
            getString(R.string.dialog_sync_link, account.getUuid()),
            new MessageDialogFragment.Button(R.string.dialog_command_sync_link_remote, R.id.SYNC_LINK_COMMAND_REMOTE, account),
            MessageDialogFragment.nullButton(android.R.string.cancel),
            new MessageDialogFragment.Button(R.string.dialog_command_sync_link_local, R.id.SYNC_LINK_COMMAND_LOCAL, account))
            .show(getParentFragmentManager(), "SYNC_LINK");
      }
      return true;
    } else if (itemId == R.id.SYNC_REMOVE_BACKEND_COMMAND) {
      String syncAccountName = syncBackendAdapter.getSyncAccountName(packedPosition);
      Bundle b = new Bundle();
      final String message = getString(R.string.dialog_confirm_sync_remove_backend, syncAccountName)
          + " " + getString(R.string.continue_confirmation);
      b.putString(ConfirmationDialogFragment.KEY_MESSAGE, message);
      b.putInt(ConfirmationDialogFragment.KEY_COMMAND_POSITIVE, R.id.SYNC_REMOVE_BACKEND_COMMAND);
      b.putInt(ConfirmationDialogFragment.KEY_POSITIVE_BUTTON_LABEL, R.string.menu_remove);
      b.putInt(ConfirmationDialogFragment.KEY_NEGATIVE_BUTTON_LABEL, android.R.string.cancel);
      b.putString(KEY_SYNC_ACCOUNT_NAME, syncAccountName);
      ConfirmationDialogFragment.newInstance(b).show(getParentFragmentManager(), "SYNC_REMOVE_BACKEND");
    }
    return super.onContextItemSelected(item);
  }

  private void requestSync(long packedPosition) {
    String syncAccountName = syncBackendAdapter.getSyncAccountName(packedPosition);
    android.accounts.Account account = GenericAccountService.getAccount(syncAccountName);
    if (ContentResolver.getIsSyncable(account, TransactionProvider.AUTHORITY) > 0) {
      Bundle bundle = new Bundle();
      bundle.putBoolean(ContentResolver.SYNC_EXTRAS_MANUAL, true);
      bundle.putBoolean(ContentResolver.SYNC_EXTRAS_EXPEDITED, true);
      ContentResolver.requestSync(account,
          TransactionProvider.AUTHORITY, bundle);
    } else {
      Bundle bundle = new Bundle(1);
      bundle.putString(KEY_SYNC_ACCOUNT_NAME, syncAccountName);
      SimpleDialog.build()
          .msg("Backend is not ready to be synced")
          .pos("Activate again")
          .extra(bundle)
          .show(this, DIALOG_INACTIVE_BACKEND);
    }
  }

  public void reloadAccountList() {
    syncBackendAdapter.setAccountList(getAccountList());
    int count = syncBackendAdapter.getGroupCount();
    for (int i = 0; i < count; i++) {
      listView.collapseGroup(i);
    }
  }

  @Override
  public void onGroupExpand(int groupPosition) {
    if (!syncBackendAdapter.hasAccountMetadata(groupPosition)) {
      metadataLoadingCount++;
      if (!snackbar.isShownOrQueued()) {
        snackbar.show();
      }
      final String backendLabel = syncBackendAdapter.getBackendLabel(groupPosition);
      viewModel.accountMetadata(backendLabel).observe(getViewLifecycleOwner(), result -> {
        metadataLoadingCount--;
        if (metadataLoadingCount == 0) {
          snackbar.dismiss();
        }
        try {
          syncBackendAdapter.setAccountMetadata(groupPosition, result.getOrThrow());
        } catch (Throwable throwable) {
          ManageSyncBackends activity = (ManageSyncBackends) requireActivity();
          if (Utils.getCause(throwable) instanceof InvalidAccessTokenException) {
            activity.requestDropboxAccess(backendLabel);
          } else {
            activity.showSnackbar(throwable.getMessage(), Snackbar.LENGTH_SHORT);
          }
        }
      });
    }
  }

  @Nullable
  public Account getAccountForSync(long packedPosition) {
    return syncBackendAdapter.getAccountForSync(packedPosition);
  }

  @Override
  public boolean onResult(@NonNull String dialogTag, int which, @NonNull Bundle extras) {
    if (dialogTag.equals(DIALOG_INACTIVE_BACKEND) && which == BUTTON_POSITIVE) {
      GenericAccountService.activateSync(GenericAccountService.getAccount(extras.getString(KEY_SYNC_ACCOUNT_NAME)), prefHandler);
    }
    return false;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/SyncBackendList.kt
package org.totschnig.myexpenses.fragment

import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ExpandableListView
import android.widget.ExpandableListView.ExpandableListContextMenuInfo
import android.widget.ExpandableListView.OnChildClickListener
import android.widget.ExpandableListView.OnGroupExpandListener
import android.widget.ExpandableListView.PACKED_POSITION_TYPE_CHILD
import android.widget.ExpandableListView.getPackedPositionGroup
import android.widget.ExpandableListView.getPackedPositionType
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.evernote.android.state.State
import com.evernote.android.state.StateSaver
import com.google.android.material.snackbar.Snackbar
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.ManageSyncBackends
import org.totschnig.myexpenses.activity.SyncBackendSetupActivity
import org.totschnig.myexpenses.activity.SyncBackendSetupActivity.Companion.REQUEST_CODE_RESOLUTION
import org.totschnig.myexpenses.adapter.SyncBackendAdapter
import org.totschnig.myexpenses.databinding.SyncBackendsListBinding
import org.totschnig.myexpenses.dialog.AccountMetaDataDialogFragment
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment
import org.totschnig.myexpenses.dialog.DialogUtils
import org.totschnig.myexpenses.dialog.MessageDialogFragment
import org.totschnig.myexpenses.feature.Feature
import org.totschnig.myexpenses.model.ContribFeature
import org.totschnig.myexpenses.model.CurrencyContext
import org.totschnig.myexpenses.model2.Account
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.provider.DatabaseConstants
import org.totschnig.myexpenses.sync.BackendService
import org.totschnig.myexpenses.sync.SyncBackendProvider
import org.totschnig.myexpenses.util.licence.LicenceHandler
import org.totschnig.myexpenses.util.safeMessage
import org.totschnig.myexpenses.viewmodel.AbstractSyncBackendViewModel
import javax.inject.Inject

class SyncBackendList : Fragment(), OnGroupExpandListener,
    OnChildClickListener {
    private var _binding: SyncBackendsListBinding? = null
    private val binding get() = _binding!!
    private lateinit var syncBackendAdapter: SyncBackendAdapter
    private var metadataLoadingCount = 0
    private lateinit var viewModel: AbstractSyncBackendViewModel

    @Inject
    lateinit var prefHandler: PrefHandler

    @Inject
    lateinit var modelClass: Class<out AbstractSyncBackendViewModel>

    @Inject
    lateinit var licenceHandler: LicenceHandler

    @Inject
    lateinit var currencyContext: CurrencyContext

    @State
    var resolutionPendingForGroup = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        val appComponent = (requireActivity().application as MyApplication).appComponent
        appComponent.inject(this)
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        viewModel = ViewModelProvider(this)[modelClass]
        appComponent.inject(viewModel)
        StateSaver.restoreInstanceState(this, savedInstanceState)
    }

    override fun onResume() {
        super.onResume()
        accountList.mapNotNull { account ->
            featureForAccount(account.first)
        }.distinct().forEach {
            manageSyncBackends.requireFeature(it)
        }
    }

    private fun featureForAccount(account: String): Feature? =
        BackendService.forAccount(account).getOrNull()?.feature

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = SyncBackendsListBinding.inflate(inflater, container, false)
        syncBackendAdapter = SyncBackendAdapter(requireContext(), currencyContext.homeCurrencyString, accountList)
        binding.list.setAdapter(syncBackendAdapter)
        binding.list.emptyView = binding.empty
        binding.list.setOnGroupExpandListener(this)
        binding.list.setOnChildClickListener(this)
        lifecycleScope.launchWhenStarted {
            viewModel.localAccountInfo.collect {
                syncBackendAdapter.setLocalAccountInfo(it)
            }
        }
        registerForContextMenu(binding.list)
        return binding.root
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        val packedPosition = (menuInfo as ExpandableListContextMenuInfo).packedPosition
        val commandId: Int
        val titleId: Int
        val isSyncAvailable = licenceHandler.hasTrialAccessTo(ContribFeature.SYNCHRONIZATION)
        if (getPackedPositionType(packedPosition) == PACKED_POSITION_TYPE_CHILD) {
            if (isSyncAvailable) {
                when (syncBackendAdapter.getSyncState(packedPosition)) {
                    SyncBackendAdapter.SyncState.SYNCED_TO_THIS -> {
                        commandId = R.id.SYNC_UNLINK_COMMAND
                        titleId = R.string.menu_sync_unlink
                    }
                    SyncBackendAdapter.SyncState.UNSYNCED -> {
                        commandId = R.id.SYNC_LINK_COMMAND
                        titleId = R.string.menu_sync_link
                    }
                    SyncBackendAdapter.SyncState.SYNCED_TO_OTHER -> {
                        commandId = R.id.SYNCED_TO_OTHER_COMMAND
                        titleId = R.string.menu_sync_link
                    }
                    SyncBackendAdapter.SyncState.UNKNOWN -> {
                        commandId = R.id.SYNC_DOWNLOAD_COMMAND
                        titleId = R.string.menu_sync_download
                    }
                    else -> {
                        commandId = 0
                        titleId = 0
                    }
                }
                if (commandId != 0) menu.add(Menu.NONE, commandId, 0, titleId)
            }
        } else {
            if (isSyncAvailable) {
                menu.add(Menu.NONE, R.id.SYNC_COMMAND, 0, R.string.menu_sync_now)
            }
            menu.add(Menu.NONE, R.id.SYNC_REMOVE_BACKEND_COMMAND, 0, R.string.remove)
            if (syncBackendAdapter.isEncrypted(packedPosition)) {
                menu.add(Menu.NONE, R.id.SHOW_PASSWORD_COMMAND, 0, R.string.input_label_passphrase)
            }
            if (
                BackendService.forAccount(
                    syncBackendAdapter.getBackendLabel(
                        getPackedPositionGroup(packedPosition)
                    )
                ).getOrNull()?.supportsReconfiguration == true
            ) {
                menu.add(Menu.NONE, R.id.RECONFIGURE_COMMAND, 0, R.string.menu_reconfigure)
            }
        }
        super.onCreateContextMenu(menu, v, menuInfo)
    }

    private val accountList: List<Pair<String, Boolean>>
        get() = viewModel.getAccounts(requireActivity())

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val packedPosition = (item.menuInfo as ExpandableListContextMenuInfo).packedPosition
        val itemId = item.itemId
        manageSyncBackends.trackCommand(itemId)
        when (itemId) {
            R.id.SYNC_COMMAND -> {
                requestSync(packedPosition)
                return true
            }
            R.id.SYNC_UNLINK_COMMAND -> {
                val accountForSync = getAccountForSync(packedPosition)
                if (accountForSync != null) {
                    DialogUtils.showSyncUnlinkConfirmationDialog(
                        manageSyncBackends,
                        accountForSync.syncAccountName, accountForSync.uuid
                    )
                }
                return true
            }
            R.id.SYNCED_TO_OTHER_COMMAND -> {
                val account = getAccountForSync(packedPosition)
                if (account != null) {
                    manageSyncBackends.showMessage(
                        getString(R.string.dialog_synced_to_other, account.uuid)
                    )
                }
                return true
            }
            R.id.SYNC_LINK_COMMAND -> {
                val account = getAccountForSync(packedPosition)
                if (account != null) {
                    MessageDialogFragment.newInstance(
                        getString(R.string.menu_sync_link),
                        getString(R.string.dialog_sync_link, account.uuid),
                        MessageDialogFragment.Button(
                            R.string.dialog_command_sync_link_remote,
                            R.id.SYNC_LINK_COMMAND_REMOTE,
                            account
                        ),
                        MessageDialogFragment.nullButton(android.R.string.cancel),
                        MessageDialogFragment.Button(
                            R.string.dialog_command_sync_link_local,
                            R.id.SYNC_LINK_COMMAND_LOCAL,
                            account
                        )
                    )
                        .show(parentFragmentManager, "SYNC_LINK")
                }
                return true
            }
            R.id.SYNC_REMOVE_BACKEND_COMMAND -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                    val syncAccountName = syncBackendAdapter.getSyncAccountName(packedPosition)
                    ConfirmationDialogFragment.newInstance(Bundle().apply {
                        putString(
                            ConfirmationDialogFragment.KEY_MESSAGE,
                            (getString(R.string.dialog_confirm_sync_remove_backend, syncAccountName)
                                    + " " + getString(R.string.continue_confirmation))
                        )
                        putInt(
                            ConfirmationDialogFragment.KEY_COMMAND_POSITIVE,
                            R.id.SYNC_REMOVE_BACKEND_COMMAND
                        )
                        putInt(
                            ConfirmationDialogFragment.KEY_POSITIVE_BUTTON_LABEL,
                            R.string.remove
                        )
                        putString(DatabaseConstants.KEY_SYNC_ACCOUNT_NAME, syncAccountName)
                    })
                        .show(parentFragmentManager, "SYNC_REMOVE_BACKEND")
                } else {
                    manageSyncBackends.showSnackBar("Account can be deleted from Android System Settings")
                }
                return true
            }
            R.id.SHOW_PASSWORD_COMMAND -> {
                viewModel.loadPassword(syncBackendAdapter.getSyncAccountName(packedPosition))
                    .observe(this) {
                        manageSyncBackends.showDismissibleSnackBar(
                            it ?: "Could not retrieve passphrase"
                        )
                    }
                return true
            }
            R.id.RECONFIGURE_COMMAND -> {
                manageSyncBackends.reconfigure(syncBackendAdapter.getSyncAccountName(packedPosition))
                return true
            }
        }
        return super.onContextItemSelected(item)
    }

    private fun requestSync(packedPosition: Long) {
        val syncAccountName = syncBackendAdapter.getSyncAccountName(packedPosition)
        manageSyncBackends.requestSync(syncAccountName)
    }

    fun reloadAccountList() {
        syncBackendAdapter.setAccountList(accountList)
        val count = syncBackendAdapter.groupCount
        for (i in 0 until count) {
            binding.list.collapseGroup(i)
        }
    }

    private val manageSyncBackends get() = requireActivity() as ManageSyncBackends

    override fun onGroupExpand(groupPosition: Int) {
        if (!syncBackendAdapter.hasAccountMetadata(groupPosition)) {
            metadataLoadingCount++
            (requireActivity() as SyncBackendSetupActivity).showLoadingSnackBar()
            val backendLabel = syncBackendAdapter.getBackendLabel(groupPosition)
            viewModel.accountMetadata(backendLabel, featureForAccount(backendLabel)?.let {
                manageSyncBackends.isFeatureAvailable(it)
            } != false)?.observe(viewLifecycleOwner) { result ->
                metadataLoadingCount--
                if (metadataLoadingCount == 0) {
                    (requireActivity() as SyncBackendSetupActivity).dismissSnackBar()
                }
                result.onSuccess { list ->
                    syncBackendAdapter.setAccountMetadata(groupPosition, list)
                }.onFailure { throwable ->
                    if (handleAuthException(throwable)) {
                        resolutionPendingForGroup = groupPosition
                    } else {
                        manageSyncBackends.showSnackBar(
                            throwable.safeMessage,
                            Snackbar.LENGTH_SHORT
                        )
                    }
                }
            }
        }
    }

    private fun handleAuthException(throwable: Throwable) =
        (throwable as? SyncBackendProvider.AuthException)?.resolution?.let {
            startActivityForResult(it, REQUEST_CODE_RESOLUTION)
            true
        } ?: false

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        StateSaver.saveInstanceState(this, outState)
    }

    fun getAccountForSync(packedPosition: Long): Account? {
        return syncBackendAdapter.getAccountForSync(packedPosition)
    }

    fun syncUnlink(uuid: String) {
        viewModel.syncUnlink(uuid).observe(this) { result ->
            result.onFailure {
                manageSyncBackends.showSnackBar(it.safeMessage)
            }
        }
    }

    override fun onChildClick(
        parent: ExpandableListView?,
        v: View?,
        groupPosition: Int,
        childPosition: Int,
        id: Long
    ): Boolean {
        syncBackendAdapter.getMetaData(groupPosition, childPosition)?.let {
            AccountMetaDataDialogFragment.newInstance(it).show(parentFragmentManager, "META_DATA")
        }
        return true
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_RESOLUTION && resultCode == RESULT_OK) {
            if (resolutionPendingForGroup != -1) {
                onGroupExpand(resolutionPendingForGroup)
                resolutionPendingForGroup = -1
            }
        }
    }
}